### PR TITLE
add Indexl and indexr

### LINKED ISSUE #82: add `index :: forall a f. Foldable => f a -> Int -> Maybe a`
If you want to get first element from some foldable you would do `find (const true)`  but if we had `index` we could grab nth value from foldable and the expression above could be written like `index 0` which is nicer I think. do others think that it would be a nice addition? if not why?

### PR DESCRIPTION
**Pull request by safareli**
fixes #82

### CODE REVIEW

**Comment by safareli** [comment]
btw, `head = indexl 0` and `last = indexr 0`

File: src/Data/Foldable.purs [hunk 1]
```diff
@@ -325,6 +327,30 @@ elem = any <<< (==)
...
+      _ ->
+        if cursor.pos == idx
+          then { elem: Just elem, pos: cursor.pos }
+          else { pos: cursor.pos + 1, elem: cursor.elem }
+
+-- | Try to get nth element from the right in a data structure
+indexr :: forall a f. Foldable f => Int -> f a -> Maybe a
+indexr idx = _.elem <<< foldr go { elem: Nothing, pos: 0 }
+  where
+  go elem cursor =
```
Comment (safareli) [inline src/Data/Foldable.purs]: I could create non-exported indexGo function and use it in in both places (in one I would need to flip it)

**Comment by safareli** [comment]
last commit changed output from:
```js
var go = function (v) {
    return function (v1) {
        if (v.pos === idx) {
            return {
                pos: v.pos + 1 | 0,
                elem: new Data_Maybe.Just(v1)
            };
        };
        var $119 = {};
        for (var $120 in v) {
            if ({}.hasOwnProperty.call(v, $120)) {
                $119[$120] = v[$120];
            };
        };
        $119.pos = v.pos + 1 | 0;
        return $119;
    };
};
```
to
```js
var go = function (cursor) {
    return function (elem1) {
        if (cursor.elem instanceof Data_Maybe.Just) {
            return cursor;
        };
        var $115 = cursor.pos === idx;
        if ($115) {
            return {
                elem: new Data_Maybe.Just(elem1),
                pos: cursor.pos
            };
        };
        return {
            pos: cursor.pos + 1 | 0,
            elem: cursor.elem
        };
    };
};
```

**Comment by safareli** [comment]
@paf31 it's ready

**Comment by garyb** [comment]
Wouldn't this sort of thing be better suited to `FoldableWithIndex`, etc? I'm not sure adding indexing to a "normal" `Foldable` is a good idea necessarily, since I don't think there's any implied order, so ideas like getting the `head` is a bit strange.

**Comment by paf31** [comment]
I do think it makes sense to add this. A `Foldable` instance just iterates its elements in some order. You can definitely observe the order by using `toList`, for example.

**Comment by safareli** [comment]
```
takel idx == toList >>> index idx 
taker idx == toList >>> \l -> index (length l - idx )
```

**Comment by matthewleon** [comment]
> I do think it makes sense to add this. A Foldable instance for a container just iterates (some or all of) its elements in some order. You can definitely observe the order by using toList, for example.

Maybe with a warning that people using this on, say, a `Map` structure should have no guarantee that the order can't change in a point release?

**Comment by paf31** [comment]
I'm not sure. It would be quite annoying, I think, for the behavior of say, an `Ord` instance to change in a minor release. I would expect this to be the same.

**Comment by safareli** [comment]
I think changing `Ord` behavior is braking change not minor. so if someone makes braking change in minor release, it's fault of the lib not of this function. Someone would still have same issue if `toList` is used for example.

**Comment by paf31** [comment]
I'm not saying that it's the fault of the function, I'm saying the order of a fold should be considered part of the observable behavior of a `Foldable` instance for the purposes of semver.

**Comment by safareli** [comment]
Is this good to go?

**Pull request merged**